File: core/src/com/xenoage/zong/core/music/slur/Slur.java
package com.xenoage.zong.core.music.slur;

import static com.xenoage.utils.CheckUtils.checkArgsNotNull;
import static com.xenoage.utils.collections.CollectionUtils.alist;
import static com.xenoage.utils.kernel.Range.range;

import java.util.ArrayList;

import lombok.Data;

import com.xenoage.utils.math.VSide;
import com.xenoage.zong.core.music.MusicElement;
import com.xenoage.zong.core.music.WaypointPosition;
import com.xenoage.zong.core.music.chord.Chord;


/**
 * A slur or tie connecting two notes,
 * represented by two {@link SlurWaypoint}s.
 * 
 * Formerly, this class was called "curved line", inspired by Wikipedia,
 * which states that "a slur is denoted with a curved line",
 * see http://en.wikipedia.org/wiki/Slur_(music) .
 * But "slur" seems to be a more intuitive naming, even if this
 * class represents both slurs and ties.
 * 
 * @author Andreas Wenger
 */
@Data public final class Slur
	implements MusicElement {

	/** Slur or tie. */
	private SlurType type;
	/** The waypoints of the slur (at least two). */
	private ArrayList<SlurWaypoint> waypoints;
	/** The vertical side of the slur, or null for default. */
	private VSide side;


	/**
	 * Creates a new {@link Slur}. 
	 * @param start  the left end point of the curved line
	 * @param stop   the right end point of the curved line
	 */
	public Slur(SlurType type, SlurWaypoint start, SlurWaypoint stop, VSide side) {
		checkArgsNotNull(type, start, stop);
		this.type = type;
		this.waypoints = alist(start, stop);
		this.side = side;
	}


	public Slur(SlurType type, ArrayList<SlurWaypoint> waypoints, VSide side) {
		if (waypoints.size() < 2)
			throw new IllegalArgumentException("At least two waypoints are needed to create a slur!");
		this.type = type;
		this.waypoints = waypoints;
		this.side = side;
	}


	public SlurWaypoint getStart() {
		return waypoints.get(0);
	}


	public SlurWaypoint getStop() {
		return waypoints.get(waypoints.size() - 1);
	}


	public SlurWaypoint getWaypoint(Chord chord) {
		for (SlurWaypoint wp : waypoints) {
			if (chord == wp.getChord()) {
				return wp;
			}
		}
		throw new IllegalArgumentException("Given chord is not part of this slur.");
	}


	public WaypointPosition getWaypointPosition(Chord chord) {
		if (chord == getStart().getChord()) {
			return WaypointPosition.Start;
		}
		else if (chord == getStop().getChord()) {
			return WaypointPosition.Stop;
		}
		else {
			for (int i : range(1, waypoints.size() - 2)) {
				if (chord == waypoints.get(i).getChord()) {
					return WaypointPosition.Continue;
				}
			}
			throw new IllegalArgumentException("Given chord is not part of this slur.");
		}
	}


	/**
	 * Replaces the given chord.
	 */
	public void replaceChord(Chord oldChord, Chord newChord) {
		for (int i : range(waypoints)) {
			SlurWaypoint wp = waypoints.get(i);
			if (wp.getChord() == oldChord) {
				//try to stay on the same note index, if possible
				wp.setNoteIndex(Math.min(wp.getNoteIndex(), newChord.getNotes().size()));
				wp.setChord(newChord);
				return;
			}
		}
		throw new IllegalArgumentException("Given chord is not part of this beam");
	}


}

File: core/src/com/xenoage/zong/core/music/slur/Slur.kt
package com.xenoage.zong.core.music.slur

import com.xenoage.utils.math.VSide
import com.xenoage.utils.math.minOrNull
import com.xenoage.utils.throwEx
import com.xenoage.zong.core.music.MusicElement
import com.xenoage.zong.core.music.MusicElementType
import com.xenoage.zong.core.music.WaypointPosition
import com.xenoage.zong.core.music.chord.Chord


/**
 * A slur or tie connecting two notes, represented by two [SlurWaypoint]s.
 *
 * Formerly, this class was called "curved line", inspired by Wikipedia,
 * which states that "a slur is denoted with a curved line",
 * see http://en.wikipedia.org/wiki/Slur_(music) .
 * But "slur" seems to be a more intuitive naming, even if this
 * class represents both slurs and ties.
 */
class Slur(
		/** Slur or tie.  */
		val type: SlurType,
		/** The waypoints of the slur (at least two).  */
		val waypoints: MutableList<SlurWaypoint>,
		/** The vertical side of the slur, or null for default.  */
		val side: VSide? = null
) : MusicElement {

	init {
		check(waypoints.size >= 2, { "At least two waypoints are needed to create a slur!" })
	}

	val start: SlurWaypoint
		get() = waypoints.first()

	val stop: SlurWaypoint
		get() = waypoints.last()

	override val elementType: MusicElementType
		get() = MusicElementType.Slur

	/**
	 * Creates a new [Slur] with the given [start] (left end point)
	 * and [end] (right end point).
	 */
	constructor(type: SlurType, start: SlurWaypoint, stop: SlurWaypoint, side: VSide?) :
			this(type, mutableListOf(start, stop), side)

	/**
	 * Gets the waypoint of the given chord, or throws an exception if the given
	 * chord is not registered as a waypoint.
	 */
	fun getWaypoint(chord: Chord): SlurWaypoint =
			waypoints.first { it.chord === chord }

	/**
	 * Gets the position (start, stop, continue) of the given chord within this
	 * slur, or throws an exception if the chord is not registered as a waypoint.
	 */
	fun getWaypointPosition(chord: Chord): WaypointPosition =
			when (waypoints.indexOfFirst { it.chord === chord }) {
				-1 -> throwEx("chord is not part of this slur")
				0 -> WaypointPosition.Start
				waypoints.lastIndex -> WaypointPosition.Stop
				else -> WaypointPosition.Continue
			}

	/**
	 * Replaces the given chord.
	 */
	fun replaceChord(oldChord: Chord, newChord: Chord) {
		val i = waypoints.indexOfFirst { it.chord === oldChord }
		if (i == -1) throwEx("chord is not part of this slur")
		val oldWaypoint = waypoints[i]
		oldWaypoint.parentSlur = null
		//try to stay on the same note index, if possible
		val newNoteIndex = minOrNull(oldWaypoint.noteIndex, newChord.notes.size)
		waypoints[i] = SlurWaypoint(newChord, newNoteIndex)
		waypoints[i].parentSlur = this
	}

	/**
	 * There are two types of slurs: The real slur ("Bindebogen" in German)
	 * and the tie ("Haltebogen" in German).
	 */
	enum class SlurType {
		/** Bindebogen. */
		Slur,
		/** Haltebogen. */
		Tie
	}

}
